lib/portage/util/_async/ForkProcess.py

```diff
@@ -11,27 +13,100 @@
+		# temporary duplicate of fd_pipes[0] so that sys.__stdin__ can
+		# be restored in the subprocess, in case this is needed for
+		# things like PROPERTIES=interactive support.
+		stdin_dup = None
 		try:
-			pid = os.fork()
+			stdin_fd = fd_pipes.get(0)
+			if stdin_fd is not None and stdin_fd == portage._get_stdin().fileno():
+				stdin_dup = os.dup(stdin_fd)
+				fcntl.fcntl(stdin_dup, fcntl.F_SETFD,
+					fcntl.fcntl(stdin_fd, fcntl.F_GETFD))
+				fd_pipes[0] = stdin_dup
+			self._proc = multiprocessing.Process(target=self._bootstrap, args=(fd_pipes,))
+			self._proc.start()
+		finally:
+			if stdin_dup is not None:
+				os.close(stdin_dup)
+
+		if self._use_proc_sentinel(self._proc):
+			self.scheduler.add_reader(self._proc.sentinel, self._proc_sentinel_exit, self._proc)
+
+		return [self._proc.pid]
+
+	def _use_proc_sentinel(self, proc):
+		"""
+		Use sentinel for versions of python that support the close
+		method, since a premature call to self._proc.close() may
+		raise ValueError. It's not safe to call remove_reader on
+		proc.sentinel unless the close method is supported, since
+		otherwise it's not possible to know when the corresponding
+		file descriptor has been closed and then reallocated to a
+		concurrent coroutine.
 
-			if pid != 0:
-				if not isinstance(pid, int):
-					raise AssertionError(
-						"fork returned non-integer: %s" % (repr(pid),))
-				return [pid]
+		@param proc: process instance
+		@type proc: multiprocessing.Process
+		@rtype: bool
+		@returns: True if multiprocessing.Process sentinal should be used
+		"""
+		return hasattr(proc, 'close')
+
+	def _cancel(self):
+		if self._proc is None:
+			super(ForkProcess, self)._cancel()
+		else:
+			self._proc.terminate()
+
+	def _async_wait(self):
+		# Wait for _proc_sentinel_exit when appropriate, since a premature
+		# call to self._proc.close() may raise ValueError.
+		if self._proc is None or not self._use_proc_sentinel(self._proc):
+			super(ForkProcess, self)._async_wait()
 
-			rval = 1
-			try:
+	def _async_waitpid(self):
+		# Wait for _proc_sentinel_exit when appropriate, since a premature
+		# call to self._proc.close() may raise ValueError.
+		if self._proc is None or not self._use_proc_sentinel(self._proc):
+			super(ForkProcess, self)._async_waitpid()
 
+	def _proc_sentinel_exit(self, proc):
+		self._proc = None
+		# If multiprocessing.Process supports the close method, then
+		# access to the sentinel value should raise ValueError if the
+		# sentinel has been closed. In this case it's not safe to call
+		# remove_reader, since file descriptor has been closed and then
+		# reallocated to a concurrent coroutine. After proc is closed,
+		# most method calls and attribute accesses will raise ValueError,
+		# so we assume that it is ready to close here and that no other
+		# code will ever close it (the sentinel provides the only means
+		# to predict when it is safe to close).
+		self.scheduler.remove_reader(proc.sentinel)
+		proc.join()
```

Comment:
This blocking `proc.join()` is not optimal. We can add a timeout 0 argument, and retry if necessary.

Reply:
Added retry via `_proc_join` method.